flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/TimerHeapInternalTimer.java

```diff
@@ -0,0 +1,246 @@
+	}
+
+	@Override
+	public long getTimestamp() {
+		return timestamp;
+	}
+
+	@Override
+	public K getKey() {
+		return key;
+	}
+
+	@Override
+	public N getNamespace() {
+		return namespace;
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) {
+			return true;
+		}
+
+		if (o instanceof InternalTimer) {
+			InternalTimer<?, ?> timer = (InternalTimer<?, ?>) o;
+			return timestamp == timer.getTimestamp()
+				&& key.equals(timer.getKey())
+				&& namespace.equals(timer.getNamespace());
+		}
+
+		return false;
+	}
+
+	/**
+	 * Returns the current index of this timer in the owning timer heap.
+	 */
+	int getTimerHeapIndex() {
+		return timerHeapIndex;
+	}
+
+	/**
+	 * Sets the current index of this timer in the owning timer heap and should only be called by the managing heap.
+	 * @param timerHeapIndex the new index in the timer heap.
+	 */
+	void setTimerHeapIndex(int timerHeapIndex) {
+		this.timerHeapIndex = timerHeapIndex;
+	}
+
+	/**
+	 * This method can be called to indicate that the timer is no longer managed be a timer heap, e.g. because it as
+	 * removed.
+	 */
+	void removedFromTimerQueue() {
+		setTimerHeapIndex(NOT_MANAGED_BY_TIMER_QUEUE_INDEX);
+	}
+
+	@Override
+	public int hashCode() {
+		int result = (int) (timestamp ^ (timestamp >>> 32));
+		result = 31 * result + key.hashCode();
+		result = 31 * result + namespace.hashCode();
+		return result;
+	}
+
+	@Override
+	public String toString() {
+		return "Timer{" +
+				"timestamp=" + timestamp +
+				", key=" + key +
+				", namespace=" + namespace +
+				'}';
+	}
+
+	/**
+	 * A {@link TypeSerializer} used to serialize/deserialize a {@link TimerHeapInternalTimer}.
+	 */
+	public static class TimerSerializer<K, N> extends TypeSerializer<InternalTimer<K, N>> {
+
+		private static final long serialVersionUID = 1119562170939152304L;
+
+		private final TypeSerializer<K> keySerializer;
+
+		private final TypeSerializer<N> namespaceSerializer;
+
+		TimerSerializer(TypeSerializer<K> keySerializer, TypeSerializer<N> namespaceSerializer) {
+			this.keySerializer = keySerializer;
+			this.namespaceSerializer = namespaceSerializer;
+		}
+
+		@Override
+		public boolean isImmutableType() {
+			return false;
+		}
+
+		@Override
+		public TypeSerializer<InternalTimer<K, N>> duplicate() {
+
+			final TypeSerializer<K> keySerializerDuplicate = keySerializer.duplicate();
+			final TypeSerializer<N> namespaceSerializerDuplicate = namespaceSerializer.duplicate();
+
+			if (keySerializerDuplicate == keySerializer &&
+				namespaceSerializerDuplicate == namespaceSerializer) {
+				// all delegate serializers seem stateless, so this is also stateless.
+				return this;
+			} else {
+				// at least one delegate serializer seems to be stateful, so we return a new instance.
+				return new TimerSerializer<>(keySerializerDuplicate, namespaceSerializerDuplicate);
+			}
+		}
+
+		@Override
+		public TimerHeapInternalTimer<K, N> createInstance() {
+			return null;
```

Comment:
Maybe we should throw a `OperationNotSupportedException` with a proper exception message instead of risking a NPE somewhere else.